include/picongpu/fields/background/templates/twtsfast/EField.tpp

```diff
@@ -0,0 +1,353 @@
+                 *
+                 * RotationMatrix[-(PI / 2+phi)].(Ey,Ex) for rotating back the field-vectors.
+                 */
+                float_64 const Ey_rot = -math::sin(+phi) * Ey_Ey;
+                float_64 const Ex_rot = -math::cos(+phi) * Ey_Ex;
+
+                /* Finally, the E-field normalized to the peak amplitude. */
+                return float3_X(float_X(Ex_rot), float_X(Ey_rot), float_X(0.0));
+            }
+
+            HDINLINE float3_X EField::operator()(DataSpace<simDim> const& cellIdx, uint32_t const currentStep) const
+            {
+                float_64 const time_SI = float_64(currentStep) * dt - tdelay;
+                traits::FieldPosition<fields::CellType, FieldE> const fieldPosE;
+
+                pmacc::math::Vector<floatD_64, detail::numComponents> const eFieldPositions_SI
+                    = detail::getFieldPositions_SI(cellIdx, halfSimSize, fieldPosE(), unit_length, focus_y_SI, phi);
+
+                /* Single TWTS-Pulse */
+                switch(pol)
+                {
+                case LINEAR_X:
+                    return getTWTSEfield_Normalized<simDim>(eFieldPositions_SI, time_SI);
+
+                case LINEAR_YZ:
+                    return getTWTSEfield_Normalized_Ey<simDim>(eFieldPositions_SI, time_SI);
+                }
+                return getTWTSEfield_Normalized<simDim>(eFieldPositions_SI, time_SI); // defensive default
+            }
+
+            /** Calculate the Ex(r,t) field here
+             *
+             * @param pos Spatial position of the target field.
+             * @param time Absolute time (SI, including all offsets and transformations) for calculating
+             *             the field */
+            HDINLINE EField::float_T EField::calcTWTSEx(float3_64 const& pos, float_64 const time) const
+            {
+                using complex_T = pmacc::math::Complex<float_T>;
+                using complex_64 = pmacc::math::Complex<float_64>;
+                /* Unit of speed */
+                float_64 const UNIT_SPEED = SI::SPEED_OF_LIGHT_SI;
+                /* Unit of time */
+                float_64 const UNIT_TIME = SI::DELTA_T_SI;
+                /* Unit of length */
+                float_64 const UNIT_LENGTH = UNIT_TIME * UNIT_SPEED;
+
+                /* Propagation speed of overlap normalized to the speed of light [Default: beta0=1.0] */
+                float_T const beta0 = float_T(beta_0);
+                /* If phi < 0 the formulas below are not directly applicable.
+                 * Instead phi is taken positive, but the entire pulse rotated by 180 deg around the
+                 * z-axis of the coordinate system in this function.
+                 */
+                float_T const phiReal = float_T(math::abs(phi));
+                float_T const alphaTilt
+                    = math::atan2(float_T(1.0) - beta0 * math::cos(phiReal), beta0 * math::sin(phiReal));
```

Comment:
please use  https://github.com/ComputationalRadiationPhysics/picongpu/blob/f6f15f7e4ff96f2ec3fc7630b7ba021b8ff83a9b/include/pmacc/algorithms/math/defines/trigo.hpp#L36
for better performance.

```C++
float_T sinVal;
float_T cosVal;
pmacc::math::sincos(phiReal, sinVal, cosVal); 
```